{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "97903fe3",
   "metadata": {},
   "source": [
    "# Call me! Crash course de llamado de variantes genómicas usando datos de sequenciación masiva\n",
    "\n",
    "## Cómo filtrar archivos VCF, python style"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f2d795f",
   "metadata": {},
   "source": [
    "Al igual que hemos hecho los demás días, lo primero es preparar el escenario, vamos a emplear las bibliotecas [Pandas](https://pandas.pydata.org/) y [Seaborn](https://seaborn.pydata.org/) para explorar nuestros archivos VCF.\n",
    "\n",
    "Para instalar pandas y seaborn en tu máquina local, requieres ejecutar los siguientes comandos:\n",
    "\n",
    "<pre>\n",
    "apt-get install python3-pip\n",
    "pip3 install pandas\n",
    "pip3 install seaborn\n",
    "pip3 install matplotlib\n",
    "pip3 install pil\n",
    "pip3 install wordcloud\n",
    "pip3 install IPython\n",
    "</pre>\n",
    "\n",
    "\n",
    "Recuerda que mientras estés en ATGlabs, los comandos anteriores no son necesarios"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdcc3e8c",
   "metadata": {},
   "source": [
    "### 1. Cargamos nuestras bibliotecas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6efe0399",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from PIL import Image\n",
    "from wordcloud import WordCloud\n",
    "from wordcloud import STOPWORDS\n",
    "from wordcloud import ImageColorGenerator\n",
    "from IPython.display import HTML"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aac3cca",
   "metadata": {},
   "source": [
    "### 2. Definimos el nombre de la muestra que nos tocó procesar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8c21765",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_name = \"S1\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d82ff770",
   "metadata": {},
   "source": [
    "### 3. Leemos nuestro archivo VCF con la función `read_csv()` de pandas y obtenemos un elemento al que llamaremos *dataframe*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb77d79a",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_file = sample_name + \"_snpeff_dbsnp_clinvar.vcf\"\n",
    "df = pd.read_csv(vcf_file,\n",
    "                 sep=\"\\t\",\n",
    "                 comment=\"#\",\n",
    "                 dtype=\"str\",\n",
    "                 header=None,\n",
    "                 names=[\"chr\",\"pos\",\"id\",\"ref\",\"alt\",\"qual\",\"filter\",\"info\",\"format\",sample_name])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f00d838f",
   "metadata": {},
   "source": [
    "#### El campo 8 de nuestro VCF está lleno de información que si bien es útil, no podemos procesarla toda de forma simultanea y tampoco podemos hacer sentido de esa información sin transformarla a algo más amigable\n",
    "\n",
    "#### Tomemos por ejemplo la primera fila de nuestro VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be425767",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"info\"][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fb27f97",
   "metadata": {},
   "source": [
    "### 4. Creamos una función para transformar la estructura del campo 8 del VCF a un `diccionario` de python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95e83591",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_var_info(info_str):\n",
    "    info_list = info_str.split(\";\")\n",
    "    info_dict = {}\n",
    "    for element in info_list:\n",
    "        if(\"=\" in element):\n",
    "            element_key   = element.split(\"=\")[0]\n",
    "            element_value = element.split(\"=\")[1]\n",
    "            info_dict[element_key] = element_value\n",
    "    return info_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5362a31",
   "metadata": {},
   "source": [
    "#### 4.1. Y transformamos nuestro campo 8 en algo menos agresivo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10455d71",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_df = df.copy()\n",
    "vcf_df[\"info\"] = vcf_df[\"info\"].apply(lambda x: get_var_info(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ba3765e",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_df[\"info\"][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "008a2ad2",
   "metadata": {},
   "source": [
    "### 5. Creamos una función para obtener el campo \"COMMON\" de nuestras variantes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "809ffc05",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_common_stat(info_dict):\n",
    "    if(\"COMMON\" in info_dict):\n",
    "        common_stat = info_dict[\"COMMON\"].split(\",\")\n",
    "    else:\n",
    "        common_stat = False\n",
    "    return common_stat"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ea46904",
   "metadata": {},
   "source": [
    "#### 5.1. Y agregamos esa columna a nuestro dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c8542ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_df[\"common\"] = vcf_df[\"info\"].apply(lambda x: get_common_stat(x))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb8b2c58",
   "metadata": {},
   "source": [
    "### 6. Creamos una función para obtener el significado clínico de nuestras variantes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd792499",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_clnsig_list(info_dict):\n",
    "    if(\"CLNSIG\" in info_dict):\n",
    "        clnsig_list = info_dict[\"CLNSIG\"].split(\",\")\n",
    "    else:\n",
    "        clnsig_list = False\n",
    "    return clnsig_list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39e4e8e9",
   "metadata": {},
   "source": [
    "#### 6.1. Y agregamos esa columna a nuestro dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ae533ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_df[\"clnsig\"] = vcf_df[\"info\"].apply(lambda x: get_clnsig_list(x))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5380551",
   "metadata": {},
   "source": [
    "### 7. Creamos una función para obtener los fenotipos asociados a nuestras variantes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2899554",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_clndn_list(info_dict):\n",
    "    if(\"CLNDN\" in info_dict):\n",
    "        clndn_list = info_dict[\"CLNDN\"].split(\"|\")\n",
    "    else:\n",
    "        clndn_list = False\n",
    "    return clndn_list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "903777a9",
   "metadata": {},
   "source": [
    "#### 7.1. Y agregamos esa columna a nuestro dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7c61bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_df[\"clndn\"] = vcf_df[\"info\"].apply(lambda x: get_clndn_list(x))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed67beca",
   "metadata": {},
   "source": [
    "### 8. Creamos una función adicional para obtener el campo \"ANN\" de nuestras variantes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1e36af5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ann_list(info_dict):\n",
    "    if(\"ANN\" in info_dict):\n",
    "        ann_list = info_dict[\"ANN\"].split(\",\")\n",
    "    else:\n",
    "        ann_list = False\n",
    "    return ann_list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9819d8fd",
   "metadata": {},
   "source": [
    "#### 8.1. Y agregamos esa columna a nuestro dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36f414e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_df[\"ann\"] = vcf_df[\"info\"].apply(lambda x: get_ann_list(x))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9059fa7c",
   "metadata": {},
   "source": [
    "### 9. Creamos una función adicional para agregar links a ensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf556e49",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ensembl_link(rs_str):\n",
    "    if \"rs\" in rs_str:\n",
    "        url_str = \"<a href=\\\"https://www.ensembl.org/homo_sapiens/Variation/Summary?v=\" + rs_str + \"\\\" target=\\\"_blank\\\">\"+rs_str+\"</a>\"\n",
    "    elif \".\" in rs_str:\n",
    "        url_str = \".\"\n",
    "    else:\n",
    "        url_str = False\n",
    "    return url_str"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08508320",
   "metadata": {},
   "source": [
    "#### 9.1. Y agregamos esa columna a nuestro dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebbdbd3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_df[\"id\"] = vcf_df[\"id\"].apply(lambda x: get_ensembl_link(x))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a15022f2",
   "metadata": {},
   "source": [
    "### 10. Limpiamos nuestro dataframe y nos quedamos con solo algunas columnas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae20c7f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_df = vcf_df[[\"chr\",\"pos\",\"id\",\"ref\",\"alt\",\"common\",\"ann\",\"clnsig\",\"clndn\",sample_name]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a405e5f",
   "metadata": {},
   "source": [
    "#### 10.1. La columna de anotaciones sigue teniendo un output algo agresivo\n",
    "\n",
    "#### Tomemos por ejemplo la primera fila de nuestro VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e118fe69",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_df[\"ann\"][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5653294c",
   "metadata": {},
   "source": [
    "### 11. Convertimos las anotaciones del campo `ann` a un formato más amigable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c054e33",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_df = vcf_df.explode(\"ann\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61da6863",
   "metadata": {},
   "source": [
    "#### 11.1. Construimos un nuevo *dataframe* y agregamos más columnas a dicho elemento a partir de la información de las anotaciones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2895d321",
   "metadata": {},
   "outputs": [],
   "source": [
    "ann_names = {0:\"allele\",1:\"effect\",2:\"impact\",3:\"gene_name\",\n",
    "             4:\"gene_id\",5:\"feature_type\",6:\"feature_id\",7:\"feature_biotype\",\n",
    "             8:\"rank\",9:\"hgvs_coding\",10:\"hgvs_protein\",11:\"cdna_pos\",\n",
    "             12:\"cds_pos\",13:\"prot_pos\",14:\"distance\",15:\"messages\"}\n",
    "ann_df = vcf_df.copy()\n",
    "ann_df = ann_df[\"ann\"].str.split(\"|\",expand=True).rename(ann_names,axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81cfd347",
   "metadata": {},
   "source": [
    "#### 11.2. Examinamos rapidamente el resultado"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5270cfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "ann_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b458224a",
   "metadata": {},
   "source": [
    "### 12. Convertimos las anotaciones del campo `sample_name` a un formato más amigable\n",
    "\n",
    "#### 12.1. Para ello, construimos un nuevo *dataframe* y agregamos más columnas a dicho elemento a partir de la información del genotipo, profundidades, calidades y likelihoods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53327739",
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_names ={0:\"genotype\",1:\"allele_depth\",2:\"depth_coverage\",3:\"genotype_quality\",4:\"phred_likelihood\"}\n",
    "gt_df = vcf_df.copy()\n",
    "gt_df = gt_df[sample_name].str.split(\":\",expand=True).rename(gt_names,axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1ec9e43",
   "metadata": {},
   "source": [
    "### 13. Integramos todo en un sólo *dataframe*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30292d55",
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_df = pd.concat([vcf_df.drop([\"ann\",sample_name],axis=1),ann_df,gt_df],axis=1)\n",
    "clean_df[\"pos\"] = clean_df[\"pos\"].astype(int)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28cb84db",
   "metadata": {},
   "source": [
    "#### 13.1. Examinamos rapidamente el resultado"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a8da179",
   "metadata": {},
   "outputs": [],
   "source": [
    "HTML(clean_df.head().to_html(escape=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23e5b541",
   "metadata": {},
   "source": [
    "### 14. Y entonces podemos filtrar por genes, genotipos, cromosomas, etc."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4db55d0",
   "metadata": {},
   "source": [
    "#### 14.1. Distribución de genotipos a lo largo de los cromosomas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb64fdcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(18,6))\n",
    "ax = sns.stripplot(data=clean_df, x=\"chr\", y=\"pos\", hue=\"genotype\")\n",
    "ax.set(ylabel=\"\")\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f6f7ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "hom_gt = [\"1/1\"]\n",
    "hom_df = clean_df.copy()\n",
    "hom_df = hom_df[hom_df[\"genotype\"].isin(hom_gt)]\n",
    "sns.histplot(hom_df[\"chr\"]);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b98a889",
   "metadata": {},
   "outputs": [],
   "source": [
    "het_gt = [\"0/1\",\"1/2\"]\n",
    "het_df = clean_df.copy()\n",
    "het_df = het_df[het_df[\"genotype\"].isin(het_gt)]\n",
    "sns.histplot(het_df[\"chr\"]);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1548dd67",
   "metadata": {},
   "source": [
    "#### 14.2. Genes sobrerepresentados en el set de variantes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f5352ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_string = clean_df[\"gene_name\"].to_list()\n",
    "gene_string = str(gene_string).replace(\"\\n\",\"\").replace(\",\",\"\").replace(\"'\",\"\").strip()\n",
    "wordcloud = WordCloud(background_color=\"white\",width=1000, height=500).generate(gene_string)\n",
    "plt.figure(figsize=(20,10))\n",
    "plt.imshow(wordcloud);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f048673d",
   "metadata": {},
   "source": [
    "#### 14.3. Variantes que pertenecen unicamente a un conjunto de genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e14970ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = [\"ABCA1\",\"ABCB4\",\"AC053503.6\",\"ADAMTS13\",\"ALMS1\",\"ASCL1\",\"ATP7B\",\"BCHE\",\"CDKN1B\",\"CEP290\",\"COL4A4\",\"COL6A3\",\"DES\",\"EIF3J\",\"EOMES\",\"F13A1\",\"F13B\",\"FBN1\",\"FGA\",\"FTH1P1\",\"GALC\",\"GALNS\",\"HCN4\",\"IFNGR1\",\"IL21R\",\"KCNV2\",\"KLKB1\",\"NPC1\",\"PAH\",\"PCDH15\",\"PCSK9\",\"PLOD1\",\"PPP4C\",\"PRNP\",\"RELN\",\"SCN5A\",\"SLC34A3\",\"SNIP1\",\"SPATA7\",\"SPG11\",\"STOX1\",\"SUMO4\",\"TAB2\",\"TBX6\",\"TTC8\",\"TYR\",\"VPS13B\",\"ZFYVE26\",\"ZFYVE27\"]\n",
    "panel_df  = clean_df.copy()\n",
    "panel_df  = panel_df[panel_df[\"gene_name\"].isin(gene_list)]\n",
    "HTML(panel_df[[\"chr\",\"pos\",\"id\",\"ref\",\"alt\",\"genotype\",\"clnsig\",\"clndn\"]].to_html(escape=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c38eb737",
   "metadata": {},
   "source": [
    "#### 14.4. Y su interpretación clínica"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbce8250",
   "metadata": {},
   "outputs": [],
   "source": [
    "clnsig_string = panel_df[\"clnsig\"].to_list()\n",
    "clnsig_string = str(clnsig_string).replace(\"\\n\",\"\").replace(\",\",\"\").replace(\"'\",\"\").strip()\n",
    "wordcloud = WordCloud(background_color=\"white\",width=1000, height=500).generate(clnsig_string)\n",
    "plt.figure(figsize=(20,10))\n",
    "plt.imshow(wordcloud);"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
